libs/balances-and-allowances/src/updaters/BalancesAndAllowancesUpdater.ts

```diff
@@ -0,0 +1,96 @@
+const BALANCES_SWR_CONFIG = { refreshInterval: ms`28s` }
+const ALLOWANCES_SWR_CONFIG = { refreshInterval: ms`30s` }
+
+export interface BalancesAndAllowancesUpdaterProps {
+  account: string | undefined
+  chainId: SupportedChainId
+}
+export function BalancesAndAllowancesUpdater({ account, chainId }: BalancesAndAllowancesUpdaterProps) {
+  const allTokens = useAllTokens()
+
+  const setBalances = useSetAtom(balancesAtom)
+  const setAllowances = useSetAtom(allowancesState)
+  const { data: nativeTokenBalance } = useNativeTokenBalance()
+
+  const spender = GP_VAULT_RELAYER[chainId]
+
+  const tokenAddresses = useMemo(() => allTokens.map((token) => token.address), [allTokens])
+  const balanceOfParams = useMemo(() => (account ? [account] : undefined), [account])
+  const allowanceParams = useMemo(() => (account && spender ? [account, spender] : undefined), [account, spender])
+
+  const { isLoading: isBalancesLoading, data: balances } = useMultipleContractSingleData<{ balance: BigNumber }>(
+    tokenAddresses,
+    erc20Interface,
+    'balanceOf',
+    balanceOfParams,
+    MULTICALL_OPTIONS,
+    BALANCES_SWR_CONFIG
+  )
+
+  const { isLoading: isAllowancesLoading, data: allowances } = useMultipleContractSingleData<[BigNumber]>(
+    tokenAddresses,
+    erc20Interface,
+    'allowance',
+    allowanceParams,
+    MULTICALL_OPTIONS,
+    ALLOWANCES_SWR_CONFIG
+  )
+
+  // Set balances loading state
+  useEffect(() => {
+    setBalances((state) => ({ ...state, isLoading: isBalancesLoading }))
+  }, [setBalances, isBalancesLoading])
+
+  // Set allwoances loading state
+  useEffect(() => {
+    setAllowances((state) => ({ ...state, isLoading: isAllowancesLoading }))
+  }, [setAllowances, isAllowancesLoading])
+
+  // Set balances to the store
+  useEffect(() => {
+    if (!balances || !balances.length) return
+
+    const balancesState = tokenAddresses.reduce<BalancesState['values']>((acc, address, index) => {
+      acc[address.toLowerCase()] = balances[index]?.balance
+      return acc
+    }, {})
+
+    // Add native token balance to the store as well
+    const nativeToken = NATIVE_CURRENCY_BUY_TOKEN[chainId]
+    const nativeBalanceState = nativeTokenBalance ? { [nativeToken.address.toLowerCase()]: nativeTokenBalance } : {}
+
+    setBalances({ isLoading: false, values: { ...balancesState, ...nativeBalanceState } })
+  }, [balances, tokenAddresses, setBalances, nativeTokenBalance, chainId])
+
+  // Set allowances to the store
+  useEffect(() => {
+    if (!allowances || !allowances.length) return
+
+    const allowancesState = tokenAddresses.reduce<AllowancesState['values']>((acc, address, index) => {
+      acc[address.toLowerCase()] = allowances[index]?.[0]
+      return acc
+    }, {})
+
+    setAllowances({ isLoading: false, values: allowancesState })
+  }, [allowances, tokenAddresses, setAllowances])
+
+  return null
```

Comment:
Yeah, it looks very similar. The first difference is that we add `nativeTokenBalance` to the balances atom.
Another one is the result of multicall:

```
// Balances
acc[address.toLowerCase()] = balances[index]?.balance

// Allowances
acc[address.toLowerCase()] = allowances[index]?.[0]
```

If I extract the common logic into a hook/function, it will be a bit complicated due to several differences.
After weighing the pros and cons, it seems to me that it is better to have a little duplicated code in favor of readability. Anyway, if you have a strong opinion I can refactor it, no problem